### PR TITLE
Array tests!

### PR DESCRIPTION
**Pull request by amosr**
! @tranma @jystic 
/jury approved @jystic @tranma

### CODE REVIEW

File: icicle-compiler/test/Icicle/Test/Foreign/Utils.hs [hunk 1]
```diff
@@ -65,8 +65,8 @@ instance Show Input where
 
 instance Arbitrary Ty where
   arbitrary = oneof_sized
-    (fmap pure [IDouble, IInt, IError, IBool, ITime, IString, IUnit])
-    [IArray <$> arbitrary]
+    (fmap pure [ITime])
+    [] -- [IArray <$> arbitrary]
```
Comment (tranma) [inline icicle-compiler/test/Icicle/Test/Foreign/Utils.hs]: ?

File: icicle-compiler/test/Icicle/Test/Foreign/Utils.hs [hunk 2]
```diff
@@ -107,89 +107,63 @@ seaOfType ty = case ty of
...
+  poke (wordPtrToPtr ptrRef) ptr
+
+allocValuePtr :: Mempool.Mempool -> Val -> IO Argument
+allocValuePtr mempool v = do
+  ptr <- Mempool.allocBytes mempool 8
+  pokeVal mempool (ptrToWordPtr ptr) v
+  return $ argPtr ptr
+
+allocInput :: Mempool.Mempool -> Input -> IO Argument
+allocInput mempool (Input _ vs) = allocValuePtr mempool (VArray vs)
```
Comment (tranma) [inline icicle-compiler/test/Icicle/Test/Foreign/Utils.hs]: this looks good

**Comment by jystic** [comment]
👍  Looking good

File: icicle-compiler/test/Icicle/Test/Foreign/Array.hs [hunk 3]
```diff
@@ -251,37 +245,33 @@ univalue ty = case ty of
...
-  runWith input $ \wptr -> runRight $ do
-    let ptr = wordPtrToPtr wptr
-    src <- readLibraryWith ["-DICICLE_ASSERT=1"] fs
-    f   <- function src fun retBool
-    r   <- liftIO $ f $ args ptr
-    -- arr <- liftIO $ peekArr ty wptr
-    -- return $ counterexample ("array after test:\n" <> show arr)
+  bracket (liftIO Mempool.create) (liftIO . Mempool.free) $ \pool -> runRight $ do
+  -- Cache the output, because we're likely to reuse the same type in another test 
+  bracketEitherT' (compileLibrary CacheLibrary ["-DICICLE_ASSERT=1"] fs) releaseLibrary $ \lib -> runRight $ do
```
Comment (amosr) [inline icicle-compiler/test/Icicle/Test/Foreign/Array.hs]: Is it OK to use CacheLibrary here? There are only seven base types, so most of the time there should be relatively few instances of the C code, but nested arrays mean there could be lots in the worst case...

Comment (tranma) [inline icicle-compiler/test/Icicle/Test/Foreign/Array.hs]: I think it's alright, most of the tests are on the base types

Comment (jystic) [inline icicle-compiler/test/Icicle/Test/Foreign/Array.hs]: does it make much difference to runtime?

Comment (amosr) [inline icicle-compiler/test/Icicle/Test/Foreign/Array.hs]: pretty significant: I think roughly 1m vs 10m

**Comment by tranma** [comment]
👍

**Pull request merged**